### PR TITLE
Adding specs

### PR DESCRIPTION
**Pull request by bpurinton**
This project starts from a blank app and builds up an industrial grade Photogram. These tests aim to match the functionality and look of the app at the end of the [final video for this project](https://canvas.uchicago.edu/courses/47618/pages/video-photogram-industrial-pt-5).

The [target for this project](https://pg-industrial-2-final.herokuapp.com/) has lot more going on with the interface. I am not testing for this, only what is shown in that final project video. This may create a situation where some of these tests fail because the student went beyond the videos and did try to match the target. I'm not sure how to handle this at the moment, and if I should test for the target, or for the video content. But, since the purpose of these initial tests are to make sure that a student has actually watched the video and tried the steps, I lean towards the latter (specs for the video content).

Done:

- [x] basic Devise sign in
- [x] `before_action :authenticate_user!` in ApplicationController
- [x] presence of bootstrap navbar
- [x] user account links in navbar
- [x] presence of `<div class="card">` for photo partial
- [x] ability to visit a user show page with `/USERNAME`
- [x] presence of comments within a photo card
- [x] ability to add a comment with the form

### CODE REVIEW

File: spec/features/1_basic_spec.rb [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+  end
+
+  it "shows the photos on bootstrap cards", points: 1 do
+    photo = Photo.create(image: "https://robohash.org/#{rand(9999)}", caption: "caption", owner_id: @user.id)
+
+    visit "/#{@user.username}"
+
+    expect(page).to have_selector("div[class='card']"),
+      "Expected /[USERNAME] to have <div class='card'> elements to display the photos."
+  end
```
Comment (jelaniwoods) [inline spec/features/1_basic_spec.rb:100]: What do you think about re-wording the title a bit
```suggestion
  it "displays the each of the user's photos inside a bootstrap card", points: 1 do
    photo = Photo.create(image: "https://robohash.org/#{rand(9999)}", caption: "caption", owner_id: @user.id)

    visit "/#{@user.username}"

    expect(page).to have_selector("div[class='card']"),
      "Expected /[USERNAME] to have a <div class='card'> element with the user's photo inside"
  end
```

This also doesn't really test that the photo is inside the div, it just checks that there _is_ a div with that class.

To be more accurate you could use `within`

```rb
  it "shows the photos on bootstrap cards", points: 1 do
    photo = Photo.create(image: "https://robohash.org/#{rand(9999)}", caption: "caption", owner_id: @user.id)

    visit "/#{@user.username}"
    within "div.card" do
      expect(page).to have_content(photo.caption),
        "Expected /[USERNAME] to have a <div class='card'> element with the user's photo inside"
    end
  end
```
The only issue with that is that if there isn't a div with class card the error message is slightly less readable and can't be customized.

>Capybara::ElementNotFound:
       Unable to find css "div.card"

What do you think?

File: spec/features/1_basic_spec.rb [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+  it "shows the comments under the photos on the bootstrap cards", points: 1 do
+    photo = Photo.create(image: "https://robohash.org/#{rand(9999)}", caption: "caption", owner_id: @user.id)
+    comment = Comment.create(body: "body", author_id: @user.id, photo_id: photo.id)
+
+    visit "/#{@user.username}"
+
+    expect(page).to have_selector("div[class='card'] ul[class^='list-group']", text: comment.body),
```
Comment (jelaniwoods) [inline spec/features/1_basic_spec.rb:108]: > ul[class^='list-group']

I'm unfamiliar with this syntax. What exactly does this check?

File: spec/features/1_basic_spec.rb (hunk 1 above)
Comment (jelaniwoods) [reply spec/features/1_basic_spec.rb:100]: Oh I guess this would also work

```rb
  it "shows the photos on bootstrap cards", points: 1 do
    photo = Photo.create(image: "https://robohash.org/#{rand(9999)}", caption: "caption", owner_id: @user.id)

    visit "/#{@user.username}"

    expect(page).to have_selector("div[class='card']", text: photo.caption),
      "Expected /[USERNAME] to have <div class='card'> elements to display the photos."
  end
```

so it's pretty much up to your preference.

File: spec/features/1_basic_spec.rb [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+      "Expected /[USERNAME] to have <ul class='list-group...'> bootstrap list elements under the photos to display comments."
+  end
+
+  it "has a bootstrap styled 'Create Comment' button", points: 1 do
```
Comment (jelaniwoods) [inline spec/features/1_basic_spec.rb:112]: 🤔 I wonder if calling this a button when _technically_ it's an `<input type="submit">` will cause students to use `<button>` instead?

File: spec/features/1_basic_spec.rb [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+describe "User authentication with the Devise gem" do
+  let(:user) { User.create(username: "alice", email: "alice@example.com", password: "password") }
+
+  it "allows a signed up user to sign in", points: 1 do
+    visit new_user_session_path
+
+    fill_in "Email", with: user.email
+    fill_in "Password", with: user.password
+    click_button "Log in"
```
Comment (jelaniwoods) [inline spec/features/1_basic_spec.rb:146]: Unsure about what the ending point of the video has, but [the target](https://pg-industrial.matchthetarget.com/users/sign_in) changed the text to "Sign in".

File: spec/features/1_basic_spec.rb [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+describe "New User record" do
+  before do
+    sign_in_user if user_model_exists?
+  end
+  
+  it "has a default `likes_count` of 0", points: 1 do
+    expect(@user.likes_count).to eq(0),
+      "Expected a new user to have a default `likes_count` of 0. Did you make the change in your user migration file?"
+  end
```
[the remark this answers is not in the document]
Comment (jelaniwoods) [reply spec/features/1_basic_spec.rb:11]: This technique also works for testing null constraints too.

```rb
User.column_for_attribute(:likes_count).null
# => true
User.column_for_attribute('created_at').null
# => false
```

File: spec/features/1_basic_spec.rb [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+    visit "/"
+
+    expect(page).to have_selector("a[href='/users/edit']", text: "Edit #{@user.username}"),
+      "Expected home page to have 'Edit [USERNAME]' link with the username of the signed in user."
+  end
+
+  it "has a sign out link with a DELETE request for the signed in user", points: 1 do
+    visit "/"
+
+    expect(page).to have_selector("a[href='/users/sign_out'][data-method='delete']"),
```
Comment (jelaniwoods) [inline spec/features/1_basic_spec.rb:41]: You could also check text here

```suggestion
    expect(page).to have_selector("a[href='/users/sign_out'][data-method='delete']", text: "Sign out"),
```

**Review by jelaniwoods — COMMENTED** [review]
@bpurinton I left some feedback!

**Comment by bpurinton** [comment]
> @bpurinton I left some feedback! I wasn't able to run the tests locally since I don't have a target handy for the ending point of the video. Do you have a branch somewhere that you were testing with?

@jelaniwoods Thanks for the feedback. This was a lot of tests all at once. Going through and leaving some comments. And here is my testing branch: https://github.com/appdev-projects/helper-methods-3/tree/part-3-solutions-bp-testing-specs

File: spec/features/1_basic_spec.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (bpurinton) [reply spec/features/1_basic_spec.rb:11]: Ah, thanks. I was just using the tools I was familiar with so far. Figures there's some more specific things for doing this. Regardless, see my next comment about these two model specs. (In favor of moving to other PR.)

File: spec/features/1_basic_spec.rb [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+    sign_in_user if user_model_exists?
+  end
+  
+  it "has a default `likes_count` of 0", points: 1 do
+    expect(@user.likes_count).to eq(0),
+      "Expected a new user to have a default `likes_count` of 0. Did you make the change in your user migration file?"
+  end
+
+  it "has a default `comments_count` of 0", points: 1 do
+    expect(@user.comments_count).to eq(0),
```
[the remark this answers is not in the document]
Comment (bpurinton) [reply spec/features/1_basic_spec.rb:14]: Yes, I was aware of the folders but was kind of unsure that it was necessary and how _we_ do it for the unique case of giving students feedback (versus actually testing a codebase for production). But regardless, Raghu just integration tests at this point and I realized later this was a database test. I think they

a) should be on the new potential (advanced) PR like the other Phase II projects
b) should be in a separate folder 

And of course I would make your suggested `column_for_attribute` changes.

File: spec/features/1_basic_spec.rb (hunk 1 above)
Comment (bpurinton) [reply spec/features/1_basic_spec.rb:100]: I like the second option, as it's in line with what I have in other places here and avoids that `div.card` error message (although Raghu does show and use the Emmett abbreviation stuff by this point in the videos).

**Comment by jelaniwoods** [comment]
Other notes about potential future test:

> User model check for database constraint on username uniqueness

You can test for indexes on a table like this

```rb
ActiveRecord::Base.connection.index_exists?(:users, :username)
# => true
```
You can also test for uniquess

```rb
ActiveRecord::Base.connection.index_exists?(:users, :username, unique: true)
```
Note that `unique: false` [does not do what you'd expect](https://stackoverflow.com/a/58018802).

TIL that the shoulda_matchers gem has [a nice helper method for this](https://github.com/thoughtbot/shoulda-matchers/blob/fd92175b8a3cfce804f7f460e3384e852b6fe3d5/lib/shoulda/matchers/active_record/have_db_index_matcher.rb#L94-L99)

```rb
it { should have_db_index(:username).unique }
```
You can test that username is `citext` instead of regular string
```rb
 User.column_for_attribute(:username).sql_type
# => "citext"
```

> Extensive belongs_to and has_many association accessors in User model
> Association accessor methods in a another model (maybe a simpler one like Comment that includes counter_cache option)
This is best done using the `shoulda_matchers` gem, mentioned in [the teachers manual](https://teachersmanual.firstdraft.com/how-to-write-tests#testing-association-accessors).

[Validations and enum type can also be tested with `shoulda_matchers`](https://github.com/thoughtbot/shoulda-matchers#activemodel-matchers)

> Presence and functionality of sample_data task

You can test rake tasks like this

```rb
Rails.application.load_tasks

describe "sample data task" do
  it "exists" do
    expect(Rake::Task.task_defined?(:sample_data)).to be true
  end
end
```
You can run a Rake task like this
```rb
Rake::Task["sample_data"].invoke
```
Which probably should have some kind of timeout on it in case the test is run when a student has an accidental infinite loop. I'm not sure if rspec has one build in.

**Comment by jelaniwoods** [comment]
> here is my testing branch: https://github.com/appdev-projects/helper-methods-3/tree/part-3-solutions-bp-testing-specs

@bpurinton this branch is for the helper-methods-3 repo. I thought helper-methods-3 and photogram-industrial were somewhat different. Are their ending points the same?

File: spec/features/1_basic_spec.rb (hunk 2 above)
Comment (bpurinton) [reply spec/features/1_basic_spec.rb:108]: The `^` means "starts with". So you could have `class="list-group foo bar baz"` and the spec would pass. Pretty neat trick to avoid any re-ordering issues / additions of margin top or bottom as is done in the videos here. I use this for the `nav` element in the tests here also because in this case (as opposed to that other nav tests in the helper-methods project) there are some changes to the class.

In that helper-methods case, I'm not going to go and change this now. Because that was a real "drop in the kitchen sink and leave it" example. And we explicitly say what the full class def should be in the test hint ([see here for what I'm talking about](https://github.com/appdev-projects/helper-methods-3/blob/main/spec/features/1_basic_spec.rb#L22-L27))

**Comment by bpurinton** [comment]
@jelaniwoods Oops, wrong link. Here's the testing branch: https://github.com/bpurinton-appdev/photogram-industrial/tree/bp-test-specs

File: spec/features/1_basic_spec.rb (hunk 3 above)
Comment (bpurinton) [reply spec/features/1_basic_spec.rb:112]: Yeah, the issue I had with this test was that they are using `form_with` and the [button syntax in the code is](https://github.com/bpurinton-appdev/photogram-industrial/blob/bp-test-specs/app/views/comments/_form.html.erb):

```erb
<div class="actions">
    <%= form.submit class: "btn btn-outline-secondary btn-block" %>
</div>
```

Which generates an `<input>` element. 

So, _if_ they are doing things correctly per the video, this will work. (form_with using a model). We could add to the hint?

File: spec/features/1_basic_spec.rb (hunk 4 above)
Comment (bpurinton) [reply spec/features/1_basic_spec.rb:146]: Video uses Log in... here's that issue of the video vs. target difference rearing its head. Going to discuss with Raghu.

File: spec/features/1_basic_spec.rb (hunk 6 above)
Comment (bpurinton) [reply spec/features/1_basic_spec.rb:41]: Thanks, yeah.

File: spec/features/1_basic_spec.rb (hunk 2 above)
Comment (jelaniwoods) [reply spec/features/1_basic_spec.rb:108]: > The ^ means "starts with". So you could have class="list-group foo bar baz" and the spec would pass. Pretty neat trick to avoid any re-ordering issues / additions of margin top or bottom as is done in the videos here.

Neat! If it means "starts with", I don't think it really addressing the CSS class re-ordering issue since

```html
<ul class="mt-2 list-group" /> 
```
would fail. Is that expected?

To check the presence of a css class on an html element regardless of order, I think [the `[attribute*=value]` selector does that](https://www.w3schools.com/cssref/sel_attr_contain.php).

File: spec/features/1_basic_spec.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (jelaniwoods) [reply spec/features/1_basic_spec.rb:14]: >a) should be on the new potential (advanced) PR like the other Phase II projects
b) should be in a separate folder

Sounds good to me

File: spec/features/1_basic_spec.rb (hunk 2 above)
Comment (bpurinton) [reply spec/features/1_basic_spec.rb:108]: Let's use the `*` instead of the `^` in this case anyway. And I agree about "RTEM".. not always done. I'm actually leaning towards your opinion here and going to change that helper-methods spec to use the wildcard search. Why not now rather than later when we get complaints...

File: spec/features/1_basic_spec.rb (hunk 3 above)
Comment (bpurinton) [reply spec/features/1_basic_spec.rb:112]: I know this is a bit janky, but here's my workaround:

```ruby
it "has a bootstrap styled 'Create Comment' button", points: 1 do
    photo = Photo.create(image: "https://robohash.org/#{rand(9999)}", caption: "caption", owner_id: @user.id)

    visit "/#{@user.username}"

    expect(page).to have_selector("input[class*='btn'][value='Create Comment']"),
      "Expected photo card to have 'Create Comment' button with bootstrap class='btn'. 
      This button was generated by the `form_with (model: ...)` helper method."
end
```

Comment (jelaniwoods) [reply spec/features/1_basic_spec.rb:96]: This works!

I'm not sure if I've mentioned it before (b/c I always forget about them), but [Grades has the ability to display hints](https://teachersmanual.firstdraft.com/how-to-write-tests#hints) that aren't just custom failure messages.

They should up like this
image

image

Comment (bpurinton) [reply spec/features/1_basic_spec.rb:96]: I was wondering about that! I assumed the "Expected blah blah" copy at the end of the `it` block and the `hint h(some_name)` were somehow redundant. But you're saying that adding the `hint:` will produce the little, clickable lightbulb?

This is pretty subtle though, and I think I'll just keep the "hint" in the "Expected..." copy rather than add another layer of complexity with the hint in the `en.yml` file.

**Comment by bpurinton** [comment]
@jelaniwoods I moved the model tests and your helpful notes to an [advanced PR](https://github.com/appdev-projects/photogram-industrial/pull/70)

Raghu signed off on what's here, so if there aren't any big errors we should get this out the door. This is just a quick patch to get feedback on the current DPI cohort.

The relevant commit for my changes to make the specs pass for video and target is [this one](https://github.com/appdev-projects/photogram-industrial/pull/69/commits/06b3157a8a6aa0218aa4be56f57067cbe95577fc)

File: spec/features/1_basic_spec.rb (hunk 3 above)
Comment (jelaniwoods) [reply spec/features/1_basic_spec.rb:96]: > But you're saying that adding the hint: will produce the little, clickable lightbulb?

Yes, and the "General Hints" section when you click on a failing test. Hints are most useful when there's a specific reading that introduces the concept being tested and we can link to it. It might not be the most helpful for this particular test, but something to keep in mind! Sometimes error messages are hard to read and we can't always customize them. Hints are nice because we can write whatever readable debug instructions we want and students can see them regardless of what the error message is.

**Review by jelaniwoods — APPROVED** [review]
@bpurinton looks good to me! 🚀

File: spec/features/1_basic_spec.rb (hunk 3 above)
Comment (bpurinton) [reply spec/features/1_basic_spec.rb:96]: I agree, using these hints is a nice feature. And I wonder if we can comb through all the specs and be more consistent in their usage. That's a whole project and I just made a [linear ticket for it](https://linear.app/firstdraft/issue/FIR-206/consistent-use-of-hints-in-specs). TODO.
